functionary/train/custom_datasets.py

```diff
@@ -205,12 +252,20 @@ def prepare_training_inputs_batch(
             final_prompt: the final prompt to be used,
             inputs: a dictionary containing: input_ids, attention_mask, labels. This will be used in model.forward(**inputs)
     """
-    # a dictionary mapping from token_id --> end_token
-    endtoken_2_id = get_end_token_to_token_id(tokenizer)
+    # a dictionary mapping from end_token_ --> end_token
+    prompt_template = get_prompt_template_from_tokenizer(tokenizer)
+    assistant_stop_token_ids = get_assistant_stop_token_ids(prompt_template, tokenizer)
+    assistant_prefix_tokens = get_prefix_assistant_token_ids(prompt_template, tokenizer)
```

Comment:
Actually these 2 should be separated because the purpose is different. And the prefixs are not necessarily special_tokens only, can be mixed with normal tokens, or only normal tokens, ...

Reply:
Ok, so maybe a more general `get_special_token_ids` for single special token like stop, assistant, user, etc. and keep `get_prefix_assistant_token_ids` as it is?

Reply:
Actually, **get_assistant_stop_token_ids** is diffrent from **get_special_token_ids**
Here we only use stop_tokens that are used for stop generation. For example, in template V1, there are a lot of stop_tokens (system, user, function_call, assistant) but we only use: function_call and assistant for get_assistant_stop_token_ids